Feature/audio support

Quillraven opened the pull request:
- add `AudioService` interface for audio API
- add `QueueAudioService` implementation to support one music track at a time and to queue sound requests to avoid playing the same sound multiple times within a single frame
- add some dummy playback to certain events like opening a chest, inventory UI, map background music, ...

@czyzby: Could you please have a look at the `QueueAudioService` implementation? I think loading sounds synchronously should be okay because they are short and should not take a lot of time but for music I am not sure. It would be great if the music gets loaded asynchronously and starts playing back when loaded because I think a player wouldn't notice if a music starts a few frames later than it was requested.
However, I struggle to implement that properly because imagine following scenario:
1) music playback gets requested and there is no `currentMusic`
2) music gets now loaded asynchronously
3) another music playback gets requested and the one from step 2 is not finished yet --> a second music load happens
4) music 1 now is finished loading but should not be played because the correct one is music 2. In fact the loading should have stopped for music 1 ideally.

I am not a concurrent programming expert and have no deep experience yet with coroutines in kotlin. Do you think this is an overkill? Or is it something that can easily be done?

Quillraven committed:
fix typo

Quillraven committed:
add AudioService and QueueAudioService implementation

Quillraven committed:
add sound and music assets

Quillraven committed:
add sound/music playback to some events

Quillraven committed:
add file path verification for sound/music

czyzby commented on commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt:
```diff
@@ -0,0 +1,190 @@
+   */
+  override fun playSound(soundFilePath: String, volume: Float) {
+    if (soundRequests.size >= maxSimultaneousSounds) {
+      LOG.info { "Reached maximum of '$maxSimultaneousSounds' simultaneous sounds. Request is ignored" }
```
Might be a bit too high logging level for this, seems like `debug` would be sufficient.

czyzby commented on commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt:
```diff
@@ -0,0 +1,190 @@
+  private val maxCachedSounds: Int = 50
+) : AudioService {
+
+  private val soundRequests = ObjectMap<String, SoundRequest>(maxSimultaneousSounds)
```
Is it not possible to play the same sound at the same time? I'd imagine sounds overlap would be pretty common depending on the game type. Jumping, shooting, or even more complex particles could play overlapping sounds at more or less the same time.

Quillraven replied:
From what I know if you play 2 sounds in the same frame the volume will be doubled, 3 times will be tripple the volume, etc... In my opinion this does not make sense and should be prevented to not explode the ears of the gamers ;)

czyzby replied:
@Quillraven But you have to consider overlapping sounds, e.g. when you play a sound in the middle of the previous one. You might want to prevent playing the same sound in the same exact frame, but if there's any kind of delay, I think that's a valid use case. For example, this might occur if you're firing a weapon multiple times, double-jumping or casting a spell with multiple projectiles. This will be especially noticeable in case of the longer sounds.

Quillraven replied:
That should still be possible because the `update` method should be called in each `render` loop and then the requests get processed and cleared and new ones can be added. Therefore, it should only avoid sound overlaps within the exact same frame. If it is a frame later, it will still be played.

czyzby replied:
Ah, I see.

czyzby commented on commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    val request = soundRequests[soundFilePath]
+    if (request != null) {
```
Consider adding the possibility to play the same sound multiple times with some overlap. This is pretty much expected in case of longer sounds.

czyzby commented on commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt:
```diff
@@ -0,0 +1,190 @@
+      LOG.debug { "Stop music '$currentMusicFilePath'" }
+      oldMusic.stop()
+      KtxAsync.launch {
+        assetStorage.unload<Music>(currentMusicFilePath)
```
It's not necessary to unload the music after it is finished. This will cause noticeable delays each time you want to change the theme. Let the engine users pre-load and manage which files should be loaded or not at any given time.

czyzby commented on commons/src/main/kotlin/com/github/quillraven/commons/audio/QueueAudioService.kt:
```diff
@@ -0,0 +1,190 @@
+   */
+  override fun update() {
+    if (!soundRequests.isEmpty) {
+      soundRequests.iterate { _, request, iterator ->
```
Don't do this. Pre-load the sounds and play the loaded files on demand. You're adding significant overhead by reloading the files each time you want to play a sound.

Quillraven replied:
I think you have a point for sound effects because even if I have 100s of them, the memory is most likely <10MB and noone cares if it gets loaded at startup.

For music it depends in my opinion. In my game it will not make a difference and I can also load it at startup or map change. I am thinking of adding that to the mapservice and in case someone defines a music property for the map then it will be played automatically.

I will think about that tomorrow :D

Also, I am not sure if the soundcache is even needed because assetStorage already provides that functionality? I also need to check that tomorrow again.

czyzby replied:
Yeah, `AssetStorage` can store all of your sounds.

czyzby replied:
As for the music files, it depends how many there are in the game, but normally I'd still at least preload a couple of themes (e.g. anything that has to do with the menu and the first few levels), and then gradually unload and load themes as the player progresses in the game. This load-on-demand approach will likely cause noticeable delays on mobile, especially since you're using synchronous loading.

Quillraven replied:
one question: the unloading of assetStorage is asynchronously, right? So that should not have an impact on the player experience?
Would it maybe make sense to keep the recent X music tracks in the service and if a music track falls out of the recent X tracks then it gets unloaded? That way the memory consumption will not grow endlessly. Or should this really be something that the user himself takes care of and I should always assume that it is handled by the user.
I would then still add a check that the Sound/Music is loaded by the assetStorage and if not log an error OR load it then synchronously.

czyzby replied:
Unloading is asynchronous, but not executed in parallel unless you launch the coroutine in the `asyncContext`. Keep in mind that some assets (e.g. textures) might be necessary to dispose of on the rendering thread.

>  Would it maybe make sense to keep the recent X music tracks in the service and if a music track falls out of the recent X tracks then it gets unloaded? That way the memory consumption will not grow endlessly. Or should this really be something that the user himself takes care of and I should always assume that it is handled by the user.

I think this should be configurable by the user. You might try both approaches at once - allow the users to preload some music files that are never unloaded (e.g. menu theme), and keep the others in a cache with a fixed size.

czyzby commented on core/src/main/kotlin/com/github/quillraven/quillycrawler/ashley/system/InteractSystem.kt:
```diff
@@ -79,11 +82,13 @@ class InteractSystem(
     when (entity.actionableCmp.type) {
       ActionType.EXIT -> {
         player.add(engine.createComponent(GoToNextLevelComponent::class.java))
+        audioService.playSound(SoundAssets.DROP.descriptor.fileName)
```
Perhaps a utility method in `SoundAssets` could shorten this a little bit, since you're using these as identifiers.

Quillraven replied:
I have that in different areas already because most methods seem to take a string or a file but they never take a descriptor unfortunately. 
Utility or extension methods make sense especially for Sounds. Will add that as well :)

czyzby replied:
`AssetStorage` supports fetching assets by their descriptor, so it's a matter of changing the `AudioService` if you'd prefer to use that to identify the assets.

czyzby reviewed (commented):
I don't think loading and unloading the music or sound files on every play request is a good approach. You're adding significant overhead for little gain. I think music and sound files should be pre-loaded - either all at once, or at specific points of the game (e.g. level change, map change). Unload assets only if you are sure that they are causing problems. In small 2D games having most assets loaded and ready to use will lead to a much smoother experience for the users.

czyzby commented:
You can try something like this for the sound assets to make it a little less verbose and error-prone (pseudo-code):

```kotlin
// Implement this with the enum:
interface SoundAsset {
  val descriptor: AssetDescriptor<Sound>

  fun play(service: AudioService, volume: Float = 1f) {
    service.playSound(descriptor.fileName, volume)
  }
}

// Usage:
SoundAssets.DROP.play(audioService)
```

This way you're avoiding passing raw strings as IDs.

Quillraven committed:
add StartUpScreen and remove load/unload from QueueAudioService

Quillraven commented:
as always, thank you for your feedback @czyzby. I added now a `StartUpScreen` (just a quick draft implementation) to preload assets and removed the loading/unloading from the audio service.

Also added two extensions for `SoundAssets` and `MusicAssets` to call `play` with the enum constant directly.


I am not 100% sure yet if I want to add the audio service to the map service and play a music when a specific property is set in the map. I think it does make sense and if a user doesn't want to use it he can simply not set that property. The only downside I see is that the audio service gets "hard" linked to the map service. Currently, I do feel this is not a bad design and is useful but again ... not 100% convinced yet :D

Quillraven committed:
add AudioService to AbstractGame

Quillraven committed:
add DefaultAudioService and music file path support for Tiled maps

Quillraven committed:
make DefaultMapService an object

Quillraven committed:
remove debug code and adjust to AbstractScreen audioService

Quillraven committed:
add AudioService and pause functionality

Quillraven committed:
add volume and pause/resume/stop for sounds

Quillraven committed:
update Kotlin to 1.4.31 and credits information

Quillraven committed:
update sounds

Quillraven committed:
update ambient sound volume

Quillraven committed:
add volume adjust logic and cap between 0 and 1
